src/http.cc

```diff
@@ -1910,10 +2037,14 @@ HttpStateData::httpBuildRequestHeader(HttpRequest * request,
         delete cc;
     }
 
-    /* maybe append Connection: keep-alive */
-    if (flags.keepalive) {
+    // Adds the Upgrade header if required. Must called before the Connection
+    // header is built.
+    makeUpgradeHeader(state, request, *hdr_in, *hdr_out);
+
+    if (hdr_out->has(Http::HdrType::UPGRADE))
+        hdr_out->putStr(Http::HdrType::CONNECTION, "Upgrade");
+    else if (flags.keepalive) /* maybe append Connection: keep-alive */
```

Comment:
Oh. The Connection can include a list not a single value.
If we just remove the else then my sense is that it will result to two separate Connection headers. I know this is acceptable, however do we need to make a new HttpHeader::addConnection ?

Reply:
> If this or a similar lambda is indeed all we need here, then please use a lambda instead of a single-use class. It is much easier to understand the embedded filter code than a separate class IMO.

Before replace the class with lamda lets see if we need to reuse the class (or lamda) in other cases. I will comment on SQUID-323 jira, where Eduard open a new related discussion.
 
However still disagree here. The reader needs to see that we are matching a protocol here. The exact matching algorithm is secondary for him and it is simpler for readers to put it out from here.

> Inside the lambda, please use more telling names instead of checking and p. Please also do not call .find() unless the call is necessary -- linear search through strings is not cheap. I suspect righting two if ... return statements would be better than using named booleans in this case.

I will see what I can do.

Reply:
> If this or a similar lambda is indeed all we need here, then please use a lambda instead of a single-use class...

Finally I kept MatchProtocol class. This is now used in two places so I think it is good option to avoid duplicate code.
Also I tried to make it as fast as possible.

Please see commit 56c94ca for details.

Reply:
> However still disagree here. The reader needs to see that we are matching a protocol here. The exact matching algorithm is secondary for him and it is simpler for readers to put it out from here.

You are absolutely right. However, when deciding on the best solution, we need to look at the resulting code as a _whole_, not just this line. Of course, replacing five lines with a single class name is going to be better for the reader of this method. What is going to be a lot harder is to comprehend what that separated-from-its-context class does. On the balance of things, lambda wins because ignoring those five lines is much easier than studying a separated-from-its-context class.

Said that, duplicating this lambda is worse than interpreting a stand-alone filter class. Based on that consideration (and not code readability), you were were right to keep the filter class! We may revisit this discussion later because the surrounding iteration code is still duplicated, and removing that duplication may result in this lambda idea resurfacing (and that is why I wanted to clarify why lambda may be overall better with all the other factors being equal). Until then, I am OK with keeping the filter class.

----

P.S. You seem to quote a change request A inside a change request B thread and then respond it it. Please avoid doing that. Please note that there are several open reviews right now. That means you have several sets of open change requests. Review Y may contain a comment on a change request XA in review X. That does not mean you should hijack an unrelated change request YB in review Y to continue talking about that change request XA -- find XA (in open review X) and continue the conversation there.

In this particular case:

* X -- Eduard's review: https://github.com/measurement-factory/squid/pull/16#pullrequestreview-182697628
* Y -- Alex' review: https://github.com/measurement-factory/squid/pull/16#pullrequestreview-184348766
* XB -- the lambda change request in Eduard's review: https://github.com/measurement-factory/squid/pull/16#discussion_r239999597
* YB -- this `else` change request hijacked for the lambda discussion: https://github.com/measurement-factory/squid/pull/16#discussion_r241216294

GitHub does not make it easy. Reviewers complicate things further by keeping several reviews open at the same time (mostly because one cannot add new change requests to their old (and still partially unaddressed) review -- one has to open a new review to add a new change request). It is mostly GitHub's fault, but avoiding hijacking helps reduce suffering.

Hint: If you do not see the change request you are addressing right above your comment, then you are probably hijacking the wrong change request.